bpf/kmesh/ads/include/cluster.h

```diff
@@ -219,14 +247,95 @@ static inline void *loadbalance_round_robin(struct cluster_endpoints *eps)
     return (void *)eps->ep_identity[idx];
 }
 
-static inline void *cluster_get_ep_identity_by_lb_policy(struct cluster_endpoints *eps, __u32 lb_policy)
+/* The daddr is explicitly excluded from the hash here in order to allow for
+ * backend selection to choose the same backend even on different service VIPs.
+ */
+// static __always_inline __u32 hash_from_tuple_v4(struct bpf_sock * sk)
+// {
+
+// 	BPF_LOG(INFO, CLUSTER, "sk: src_ip is:%x, src_port is:%d\n", sk->src_ip4,sk->src_port);
+// 	BPF_LOG(INFO, CLUSTER, "sk: dst_port is:%d, sk->protocol is: %x\n",sk->dst_port,sk->protocol);
+// 	return jhash_3words(sk->src_ip4,
+// 			    ((__u32) sk->dst_port << 16) | sk->src_port,
+// 			    sk->protocol, HASH_INIT4_SEED);
+// }
+
+static __always_inline __u32 map_array_get_32(__u32 *array, __u32 index, const __u32 limit)
+{
+    __u32 datum = 0;
+
+    // if (__builtin_constant_p(index) ||
+    //     !__builtin_constant_p(limit))
+    // 	__throw_build_bug();
+
+    /* LLVM tends to optimize code away that is needed for the verifier to
+     * understand dynamic map access. Input constraint is that index < limit
+     * for this util function, so we never fail here, and returned datum is
+     * always valid.
+     */
+    asm volatile("%[index] <<= 2\n\t"
+                 "if %[index] > %[limit] goto +1\n\t"
+                 "%[array] += %[index]\n\t"
+                 "%[datum] = *(u32 *)(%[array] + 0)\n\t"
+                 : [datum] "=r"(datum)
+                 : [limit] "i"(limit), [array] "r"(array), [index] "r"(index)
+                 : /* no clobbers */);
+
+    return datum;
+}
+
+static inline void *loadbalance_maglev_select_backend(struct cluster_endpoints *eps, char *name, ctx_buff_t *ctx)
+{
+    if (!eps || eps->ep_num == 0)
+        return NULL;
+
+    __u32 inner_key = 0;
+    __u32 *backend_ids;
+    __u32 index;
+    __u32 id;
+    __u32 *inner_of_maglev;
+    struct bpf_sock_ops *skops;
+    __u32 *hash_ptr;
+    __u32 hash;
+
+    inner_of_maglev = map_lookup_cluster_inner_map(name);
+    if (!inner_of_maglev) {
+        return NULL;
+    }
+    backend_ids = bpf_map_lookup_elem(inner_of_maglev, &inner_key);
+    if (!backend_ids) {
+        return NULL;
+    }
+    hash_ptr = map_lookup_lb_hash();
+    if (!hash_ptr || *hash_ptr == 0) {
+        hash = bpf_get_prandom_u32();
+    }else {
+        hash = *hash_ptr;
+        BPF_LOG(INFO, CLUSTER, "lb_policy is maglev, got a hash value:%u\n", hash);
+    }
+    index = hash % MAGLEV_TABLE_SIZE;
+    if (index >= MAGLEV_TABLE_SIZE)
```

Comment:
Will `index` is greater than `MAGLEV_TABLE_SIZE` after the `%` operation?

Reply:
In ebpf prog, add this condition to prevent load prog to kernel failed.